cirq/testing/order_tester.py

```diff
@@ -24,78 +24,65 @@
 
     def __init__(self):
         self.groups = []
 
+    def _old_python(self):
+        return sys.version_info < (3,)
+
+    def _try_comparison(self, comp):
+        result = NotImplemented
+        try:
+            result = comp()
+        except TypeError:
+            pass # will return NotImplemented
+        return result
+
+    def _do_assert(self, one, two, symbol):
+
+        condition = one or two  # bool(NotImplemented) == True
+        message = "{} is inconsistent: {!r},{!r}".format(symbol, one, two)
+
+        assert condition, message
+
+
     def _verify_ascending(self, v1, v2):
         """Verifies that (v1, v2) is a strictly ascending sequence."""
 
-        lt_1 = TryComparison(NotImplemented, lambda: v1 < v2)
-        not_lt_2 = TryComparison(NotImplemented, lambda: not v2 < v1)
-
-        assert (lt_1, not_lt_2) in [
-            (True, True),
-            (True, NotImplemented),
-            (NotImplemented, True),
-            (NotImplemented, NotImplemented)
-        ], ("__lt__ is inconsistent: {!r},{!r}".format(lt_1, not_lt_2))
-
-        gt_2 = TryComparison(NotImplemented,
-            lambda: not hasattr(v2, '__gt__') or v2 > v1)
-        not_gt_1 = TryComparison(NotImplemented,
-            lambda: not hasattr(v1, '__gt__') or not v1 > v2)
-
-        assert (gt_2, not_gt_1) in [
-            (True, True),
-            (True, NotImplemented),
-            (NotImplemented, True),
-            (NotImplemented, NotImplemented),
-                ]
-
-        # at least one strict ordering operator should work
-        assert (lt_1, gt_2) in [
-            (True, True),
-            (True, NotImplemented),
-            (NotImplemented, True),
-        ], ("__gt__ is inconsistent: {!r},{!r}".format(lt_1, gt_2))
-
-        le_1 = TryComparison(NotImplemented,
-            lambda: not hasattr(v1, '__le__') or v1 <= v2)
-        not_le_2 = TryComparison(NotImplemented,
-            lambda: not hasattr(v2, '__le__') or not v2 <= v1)
-
-        assert (le_1, not_le_2) in [
-            (True, True),
-            (True, NotImplemented),
-            (NotImplemented, True),
-            (NotImplemented, NotImplemented),
-        ], ("__le__ is inconsistent: {!r},{!r}".format(le_1, not_le_2))
-
-        ge_2 = TryComparison(NotImplemented,
-            lambda: not hasattr(v2, '__ge__') or v2 >= v1)
-        not_ge_1 = TryComparison(NotImplemented,
-            lambda: not hasattr(v1, '__ge__') or not v1 >= v2)
-
-        assert (ge_2, not_ge_1) in [
-            (True, True),
-            (True, NotImplemented),
-            (NotImplemented, True),
-            (NotImplemented, NotImplemented),
-        ], ("__ge__ is inconsistent: {!r},{!r}".format(ge_2, not_ge_1))
+        lt_1 = self._try_comparison(lambda: v1 < v2)
+        not_lt_2 = self._try_comparison(lambda: not v2 < v1)
+
+        self._do_assert(lt_1, not_lt_2, "{!r}__lt__{!r}".format(v1, v2))
```

Comment:
This is too lenient. The check should fail if the values are incomparible (i.e. throw a type error). Just `assert v1 < v2` and `assert not v2 < v1`.

Reply:
The current checks in that method are supposed to do the following (assuming we want to verify ascending order for a, b):

- for Python3, verify that strictly either a < b or b > a; also verify that a<=b,  b>=a or le / ge arent defined

- for Python2, only verify that either a<b or b>a.

But right, type error should just mean a no go. I  thought that in theory it could be possible to have type error only for some comparison options, but probably it's too exotic.

Reply:
We want both comparisons to succeed in python 2 also. Why would we be okay with a class that failed if you did b>a instead of a<b?